// relay-compiler/lib/RelayCompilerPublic.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 *  strict-local
 * @format
 */

'use strict';

var RelayJSModuleParser = require('./RelaySourceModuleParser')(require('./FindGraphQLTags').find);

module.exports = {
  ConsoleReporter: require('graphql-compiler').ConsoleReporter,
  Parser: require('./RelayParser'),
  CodeGenerator: require('./RelayCodeGenerator'),

  GraphQLCompilerContext: require('graphql-compiler').CompilerContext,

  /** @deprecated Use JSModuleParser. */
  FileIRParser: RelayJSModuleParser,

  FileWriter: require('./RelayFileWriter'),
  IRTransforms: require('./RelayIRTransforms'),
  JSModuleParser: RelayJSModuleParser,
  MultiReporter: require('graphql-compiler').MultiReporter,
  Runner: require('graphql-compiler').CodegenRunner,
  compileRelayArtifacts: require('./compileRelayArtifacts'),
  formatGeneratedModule: require('./formatGeneratedModule'),
  convertASTDocuments: require('graphql-compiler').ASTConvert.convertASTDocuments,
  transformASTSchema: require('graphql-compiler').ASTConvert.transformASTSchema
};
