// relay-compiler/lib/RelayIRTransforms.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 *  strict-local
 * @format
 */

'use strict';

// Transforms applied to the code used to process a query response.
var relaySchemaExtensions = [require('./RelayConnectionTransform').SCHEMA_EXTENSION, require('./RelayRelayDirectiveTransform').SCHEMA_EXTENSION];

// Transforms applied to both operations and fragments for both reading and
// writing from the store.
var relayCommonTransforms = [require('./RelayConnectionTransform').transform, require('./RelayViewerHandleTransform').transform, require('./RelayRelayDirectiveTransform').transform, require('./RelayMaskTransform').transform, require('./RelayDeferrableFragmentTransform').transform];

// Transforms applied to fragments used for reading data from a store
var relayFragmentTransforms = [require('./RelayFieldHandleTransform').transform, require('graphql-compiler').FlattenTransform.transformWithOptions({ flattenAbstractTypes: true }), require('graphql-compiler').SkipRedundantNodesTransform.transform];

// Transforms applied to queries/mutations/subscriptions that are used for
// fetching data from the server and parsing those responses.
var relayQueryTransforms = [require('./RelayApplyFragmentArgumentTransform').transform, require('graphql-compiler').SkipClientFieldTransform.transform, require('graphql-compiler').SkipUnreachableNodeTransform.transform, require('./RelayGenerateIDFieldTransform').transform];

// Transforms applied to the code used to process a query response.
var relayCodegenTransforms = [require('graphql-compiler').InlineFragmentsTransform.transform, require('graphql-compiler').FlattenTransform.transformWithOptions({ flattenAbstractTypes: true }), require('graphql-compiler').SkipRedundantNodesTransform.transform, require('./RelayGenerateTypeNameTransform').transform, require('graphql-compiler').FilterDirectivesTransform.transform];

// Transforms applied before printing the query sent to the server.
var relayPrintTransforms = [require('graphql-compiler').FlattenTransform.transformWithOptions({}), require('./RelayGenerateTypeNameTransform').transform, require('./RelaySkipHandleFieldTransform').transform, require('graphql-compiler').FilterDirectivesTransform.transform, require('graphql-compiler').StripUnusedVariablesTransform.transform];

module.exports = {
  commonTransforms: relayCommonTransforms,
  codegenTransforms: relayCodegenTransforms,
  fragmentTransforms: relayFragmentTransforms,
  printTransforms: relayPrintTransforms,
  queryTransforms: relayQueryTransforms,
  schemaExtensions: relaySchemaExtensions
};
