// graphql-compiler/lib/GraphQLParser.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */

'use strict';

var _extends3 = _interopRequireDefault(require('babel-runtime/helpers/extends'));

var _classCallCheck3 = _interopRequireDefault(require('babel-runtime/helpers/classCallCheck'));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var ARGUMENT_DEFINITIONS = 'argumentDefinitions';
var ARGUMENTS = 'arguments';

/**
 * @internal
 *
 * This directive is not intended for use by developers directly. To set a field
 * handle in product code use a compiler plugin.
 */
var CLIENT_FIELD = '__clientField';
var CLIENT_FIELD_HANDLE = 'handle';
var CLIENT_FIELD_KEY = 'key';
var CLIENT_FIELD_FILTERS = 'filters';

var INCLUDE = 'include';
var SKIP = 'skip';
var IF = 'if';

var GraphQLParser = function () {
  GraphQLParser.parse = function parse(schema, text, filename) {
    var _this = this;

    var ast = require('graphql').parse(new (require('graphql').Source)(text, filename));
    var nodes = [];
    // TODO T24511737 figure out if this is dangerous
    schema = require('graphql').extendSchema(schema, ast, { assumeValid: true });
    ast.definitions.forEach(function (definition) {
      if (require('./GraphQLSchemaUtils').isExecutableDefinitionAST(definition)) {
        nodes.push(_this.transform(schema, definition));
      }
    }, this);
    return nodes;
  };

  /**
   * Transforms a raw GraphQL AST into a simpler representation with type
   * information.
   */


  GraphQLParser.transform = function transform(schema, definition) {
    var _this2 = this;

    return require('./GraphQLCompilerProfiler').run('GraphQLParser.transform', function () {
      var parser = new _this2(schema, definition);
      return parser.transform();
    });
  };

  function GraphQLParser(schema, definition) {
    (0, _classCallCheck3['default'])(this, GraphQLParser);

    this._definition = definition;
    this._referencedVariableTypesByName = {};
    this._schema = schema;
  }

  /**
   * Find the definition of a field of the specified type.
   */


  GraphQLParser.prototype.getFieldDefinition = function getFieldDefinition(parentType, fieldName, fieldAST) {
    var type = require('./GraphQLSchemaUtils').getRawType(parentType);
    var isQueryType = type === this._schema.getQueryType();
    var hasTypeName = type instanceof require('graphql').GraphQLObjectType || type instanceof require('graphql').GraphQLInterfaceType || type instanceof require('graphql').GraphQLUnionType;

    var schemaFieldDef = void 0;
    if (isQueryType && fieldName === require('graphql').SchemaMetaFieldDef.name) {
      schemaFieldDef = require('graphql').SchemaMetaFieldDef;
    } else if (isQueryType && fieldName === require('graphql').TypeMetaFieldDef.name) {
      schemaFieldDef = require('graphql').TypeMetaFieldDef;
    } else if (hasTypeName && fieldName === require('graphql').TypeNameMetaFieldDef.name) {
      schemaFieldDef = require('graphql').TypeNameMetaFieldDef;
    } else if (type instanceof require('graphql').GraphQLInterfaceType || type instanceof require('graphql').GraphQLObjectType) {
      schemaFieldDef = type.getFields()[fieldName];
    }
    return schemaFieldDef;
  };

  GraphQLParser.prototype._getErrorContext = function _getErrorContext() {
    var message = 'document `' + getName(this._definition) + '`';
    if (this._definition.loc && this._definition.loc.source) {
      message += ' file: `' + this._definition.loc.source.name + '`';
    }
    return message;
  };

  GraphQLParser.prototype._recordAndVerifyVariableReference = function _recordAndVerifyVariableReference(name, usedAsType) {
    var previousType = this._referencedVariableTypesByName[name];

    if (!previousType) {
      // No previous usage, current type is strongest
      this._referencedVariableTypesByName[name] = usedAsType;
    } else if (usedAsType) {
      !(require('graphql').isTypeSubTypeOf(this._schema, usedAsType, previousType) || require('graphql').isTypeSubTypeOf(this._schema, previousType, usedAsType)) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Variable `$%s` was used in locations expecting ' + 'the conflicting types `%s` and `%s`. Source: %s.', name, previousType, usedAsType, this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;

      // If the new used type has stronger requirements, use that type as reference,
      // otherwise keep referencing the previous type
      this._referencedVariableTypesByName[name] = require('graphql').isTypeSubTypeOf(this._schema, usedAsType, previousType) ? usedAsType : previousType;
    }
  };

  GraphQLParser.prototype.transform = function transform() {
    switch (this._definition.kind) {
      case 'OperationDefinition':
        return this._transformOperation(this._definition);
      case 'FragmentDefinition':
        return this._transformFragment(this._definition);
      default:
        !false ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Unknown AST kind `%s`. Source: %s.', this._definition.kind, this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
    }
  };

  GraphQLParser.prototype._transformFragment = function _transformFragment(fragment) {
    var _this3 = this;

    var argumentDefinitions = this._buildArgumentDefinitions(fragment);
    var directives = this._transformDirectives((fragment.directives || []).filter(function (directive) {
      return getName(directive) !== ARGUMENT_DEFINITIONS;
    }));
    var type = require('graphql').assertCompositeType(require('./GraphQLSchemaUtils').getTypeFromAST(this._schema, fragment.typeCondition));
    var selections = this._transformSelections(fragment.selectionSet, type);

    var _loop = function _loop(_name) {
      if (_this3._referencedVariableTypesByName.hasOwnProperty(_name)) {
        var variableType = _this3._referencedVariableTypesByName[_name];
        var localArgument = argumentDefinitions.find(function (argDef) {
          return argDef.name === _name;
        });
        if (localArgument) {
          !(variableType == null || require('graphql').isTypeSubTypeOf(_this3._schema, localArgument.type, variableType)) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Variable `$%s` was defined as type `%s`, but used ' + 'in a location that expects type `%s`. Source: %s.', _name, localArgument.type, variableType, _this3._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
        } else {
          argumentDefinitions.push({
            kind: 'RootArgumentDefinition',
            metadata: null,
            name: _name,
            // $FlowFixMe - could be null
            type: variableType
          });
        }
      }
    };

    for (var _name in this._referencedVariableTypesByName) {
      _loop(_name);
    }
    return {
      kind: 'Fragment',
      directives: directives,
      metadata: null,
      name: getName(fragment),
      selections: selections,
      type: type,
      argumentDefinitions: argumentDefinitions
    };
  };

  /**
   * Polyfills suport for fragment variable definitions via the
   * @argumentDefinitions directive. Returns the equivalent AST
   * to the `argumentDefinitions` property on queries/mutations/etc.
   */


  GraphQLParser.prototype._buildArgumentDefinitions = function _buildArgumentDefinitions(fragment) {
    var _this4 = this;

    var variableDirectives = (fragment.directives || []).filter(function (directive) {
      return getName(directive) === ARGUMENT_DEFINITIONS;
    });
    if (!variableDirectives.length) {
      return [];
    }
    !(variableDirectives.length === 1) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Directive %s may be defined at most once on fragment ' + '`%s`. Source: %s.', ARGUMENT_DEFINITIONS, getName(fragment), this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
    var variableDirective = variableDirectives[0];
    // $FlowIssue: refining directly on `variableDirective.arguments` doesn't
    // work, below accesses all report arguments could still be null/undefined.
    var args = variableDirective.arguments;
    if (variableDirective == null || !Array.isArray(args)) {
      return [];
    }
    !args.length ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Directive %s requires arguments: remove the directive ' + 'to skip defining local variables for this fragment `%s`. Source: %s.', ARGUMENT_DEFINITIONS, getName(fragment), this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
    return args.map(function (arg) {
      var argName = getName(arg);
      var argValue = _this4._transformValue(arg.value);
      !(argValue.kind === 'Literal') ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected definition for variable `%s` to be an ' + 'object with the following shape: `{type: string, defaultValue?: ' + 'mixed}`, got `%s`. Source: %s.', argValue, _this4._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
      var value = argValue.value;
      if (Array.isArray(value) || typeof value !== 'object' || value === null || typeof value.type !== 'string') {
        throw new Error('GraphQLParser: Expected definition for variable `' + argName + '` to ' + 'be an object with the following shape: `{type: string, ' + 'defaultValue?: mixed, nonNull?: boolean, list?: boolean}`, got ' + ('`' + String(argValue) + '`. Source: ' + _this4._getErrorContext() + '.'));
      }

      var valueType = value.type;

      var unknownKeys = Object.keys(value).filter(function (key) {
        return key !== 'type' && key !== 'defaultValue' && key !== 'nonNull' && key !== 'list';
      });
      if (unknownKeys.length !== 0) {
        var unknownKeysString = '`' + unknownKeys.join('`, `') + '`';
        throw new Error('GraphQLParser: Expected definition for variable `' + argName + '` to ' + 'be an object with the following shape: `{type: string, ' + 'defaultValue?: mixed, nonNull?: boolean, list?: boolean}`, got ' + ('unknown key(s) ' + unknownKeysString + '. ') + ('Source: ' + _this4._getErrorContext() + '.'));
      }

      var typeAST = require('graphql').parseType(valueType);
      var type = require('graphql').assertInputType(require('./GraphQLSchemaUtils').getTypeFromAST(_this4._schema, typeAST));
      return {
        kind: 'LocalArgumentDefinition',
        defaultValue: value.defaultValue != null ? value.defaultValue : null,
        metadata: null,
        name: argName,
        type: type
      };
    });
  };

  GraphQLParser.prototype._transformOperation = function _transformOperation(definition) {
    var name = getName(definition);
    var argumentDefinitions = this._transformArgumentDefinitions(definition.variableDefinitions || []);
    var directives = this._transformDirectives(definition.directives || []);
    var type = void 0;
    var operation = void 0;
    switch (definition.operation) {
      case 'query':
        operation = 'query';
        type = require('graphql').assertCompositeType(this._schema.getQueryType());
        break;
      case 'mutation':
        operation = 'mutation';
        type = require('graphql').assertCompositeType(this._schema.getMutationType());
        break;
      case 'subscription':
        operation = 'subscription';
        type = require('graphql').assertCompositeType(this._schema.getSubscriptionType());
        break;
      default:
        !false ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Unknown AST kind `%s`. Source: %s.', definition.operation, this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
    }
    !definition.selectionSet ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected %s `%s` to have selections. Source: %s.', operation, name, this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
    var selections = this._transformSelections(definition.selectionSet, type);
    return {
      kind: 'Root',
      operation: operation,
      metadata: null,
      name: name,
      dependentRequests: [],
      argumentDefinitions: argumentDefinitions,
      directives: directives,
      selections: selections,
      type: type
    };
  };

  GraphQLParser.prototype._transformArgumentDefinitions = function _transformArgumentDefinitions(argumentDefinitions) {
    var _this5 = this;

    return argumentDefinitions.map(function (def) {
      var name = getName(def.variable);
      var type = require('graphql').assertInputType(require('./GraphQLSchemaUtils').getTypeFromAST(_this5._schema, def.type));
      var defaultLiteral = def.defaultValue ? _this5._transformValue(def.defaultValue) : null;
      if (_this5._referencedVariableTypesByName.hasOwnProperty(name)) {
        var variableType = _this5._referencedVariableTypesByName[name];
        !(variableType == null || require('graphql').isTypeSubTypeOf(_this5._schema, type, variableType)) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Variable `$%s` was defined as type `%s`, but used ' + 'in a location that expects type `%s`. Source: %s.', name, type, variableType, _this5._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
      }
      !(defaultLiteral === null || defaultLiteral.kind === 'Literal') ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected null or Literal default value, got: `%s`. ' + 'Source: %s.', defaultLiteral && defaultLiteral.kind, _this5._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
      return {
        kind: 'LocalArgumentDefinition',
        metadata: null,
        name: name,
        defaultValue: defaultLiteral ? defaultLiteral.value : null,
        type: type
      };
    });
  };

  GraphQLParser.prototype._transformSelections = function _transformSelections(selectionSet, parentType) {
    var _this6 = this;

    return selectionSet.selections.map(function (selection) {
      var node = void 0;
      if (selection.kind === 'Field') {
        node = _this6._transformField(selection, parentType);
      } else if (selection.kind === 'FragmentSpread') {
        node = _this6._transformFragmentSpread(selection, parentType);
      } else if (selection.kind === 'InlineFragment') {
        node = _this6._transformInlineFragment(selection, parentType);
      } else {
        !false ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Unexpected AST kind `%s`. Source: %s.', selection.kind, _this6._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
      }

      var _splitConditions2 = _this6._splitConditions(node.directives),
          conditions = _splitConditions2[0],
          directives = _splitConditions2[1];

      var conditionalNodes = applyConditions(conditions,
      // $FlowFixMe(>=0.28.0)
      [(0, _extends3['default'])({}, node, { directives: directives })]);
      !(conditionalNodes.length === 1) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected exactly one conditional node, got `%s`. ' + 'Source: %s.', conditionalNodes.length, _this6._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
      return conditionalNodes[0];
    });
  };

  GraphQLParser.prototype._transformInlineFragment = function _transformInlineFragment(fragment, parentType) {
    var typeCondition = require('graphql').assertCompositeType(fragment.typeCondition ? require('./GraphQLSchemaUtils').getTypeFromAST(this._schema, fragment.typeCondition) : parentType);
    var directives = this._transformDirectives(fragment.directives || []);
    var selections = this._transformSelections(fragment.selectionSet, typeCondition);
    return {
      kind: 'InlineFragment',
      directives: directives,
      metadata: null,
      selections: selections,
      typeCondition: typeCondition
    };
  };

  GraphQLParser.prototype._transformFragmentSpread = function _transformFragmentSpread(fragment, parentType) {
    var _this7 = this;

    var fragmentName = getName(fragment);

    var _partitionArray = partitionArray(fragment.directives || [], function (directive) {
      return getName(directive) !== ARGUMENTS;
    }),
        otherDirectives = _partitionArray[0],
        argumentDirectives = _partitionArray[1];

    !(argumentDirectives.length <= 1) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Directive %s may be used at most once in fragment ' + 'spread `...%s`. Source: %s.', ARGUMENTS, fragmentName, this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
    var args = void 0;
    if (argumentDirectives.length) {
      args = (argumentDirectives[0].arguments || []).map(function (arg) {
        var argValue = arg.value;
        !(argValue.kind === 'Variable') ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: All @arguments() args must be variables, got %s. ' + 'Source: %s.', argValue.kind, _this7._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;

        return {
          kind: 'Argument',
          metadata: null,
          name: getName(arg),
          value: _this7._transformVariable(argValue),
          type: null // TODO: can't get type until referenced fragment is defined
        };
      });
    }
    var directives = this._transformDirectives(otherDirectives);
    return {
      kind: 'FragmentSpread',
      args: args || [],
      metadata: null,
      name: fragmentName,
      directives: directives
    };
  };

  GraphQLParser.prototype._transformField = function _transformField(field, parentType) {
    var name = getName(field);
    var fieldDef = this.getFieldDefinition(parentType, name, field);

    !fieldDef ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Unknown field `%s` on type `%s`. Source: %s.', name, parentType, this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
    var alias = field.alias ? field.alias.value : null;
    var args = this._transformArguments(field.arguments || [], fieldDef.args);

    var _partitionArray2 = partitionArray(field.directives || [], function (directive) {
      return getName(directive) !== CLIENT_FIELD;
    }),
        otherDirectives = _partitionArray2[0],
        clientFieldDirectives = _partitionArray2[1];

    var directives = this._transformDirectives(otherDirectives);
    var type = require('graphql').assertOutputType(fieldDef.type);
    var handles = this._transformHandle(name, args, clientFieldDirectives);
    if (require('graphql').isLeafType(require('graphql').getNamedType(type))) {
      !(!field.selectionSet || !field.selectionSet.selections || !field.selectionSet.selections.length) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected no selections for scalar field `%s` on type ' + '`%s`. Source: %s.', name, this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
      return {
        kind: 'ScalarField',
        alias: alias,
        args: args,
        directives: directives,
        handles: handles,
        metadata: null,
        name: name,
        type: assertScalarFieldType(type)
      };
    } else {
      var selections = field.selectionSet ? this._transformSelections(field.selectionSet, type) : null;
      !(selections && selections.length) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected at least one selection for non-scalar field ' + '`%s` on type `%s`. Source: %s.', name, type, this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
      return {
        kind: 'LinkedField',
        alias: alias,
        args: args,
        directives: directives,
        handles: handles,
        metadata: null,
        name: name,
        selections: selections,
        type: type
      };
    }
  };

  GraphQLParser.prototype._transformHandle = function _transformHandle(fieldName, fieldArgs, clientFieldDirectives) {
    var _this8 = this;

    var handles = void 0;
    clientFieldDirectives.forEach(function (clientFieldDirective) {
      var handleArgument = (clientFieldDirective.arguments || []).find(function (arg) {
        return getName(arg) === CLIENT_FIELD_HANDLE;
      });
      if (handleArgument) {
        var _name2 = null;
        var key = require('./DefaultHandleKey').DEFAULT_HANDLE_KEY;
        var filters = null;
        var maybeHandle = _this8._transformValue(handleArgument.value);
        !(maybeHandle.kind === 'Literal' && typeof maybeHandle.value === 'string') ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected the %s argument to @%s to be a literal ' + 'string, got `%s` on field `%s`. Source: %s.', CLIENT_FIELD_HANDLE, CLIENT_FIELD, maybeHandle, fieldName, _this8._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
        _name2 = maybeHandle.value;

        var keyArgument = (clientFieldDirective.arguments || []).find(function (arg) {
          return getName(arg) === CLIENT_FIELD_KEY;
        });
        if (keyArgument) {
          var maybeKey = _this8._transformValue(keyArgument.value);
          !(maybeKey.kind === 'Literal' && typeof maybeKey.value === 'string') ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected %s argument to @%s to be a literal ' + 'string, got `%s` on field `%s`. Source: %s.', CLIENT_FIELD_KEY, CLIENT_FIELD, maybeKey, fieldName, _this8._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
          key = maybeKey.value;
        }
        var filtersArgument = (clientFieldDirective.arguments || []).find(function (arg) {
          return getName(arg) === CLIENT_FIELD_FILTERS;
        });
        if (filtersArgument) {
          var maybeFilters = _this8._transformValue(filtersArgument.value);
          !(maybeFilters.kind === 'Literal' && Array.isArray(maybeFilters.value) && maybeFilters.value.every(function (filter) {
            return fieldArgs.some(function (fieldArg) {
              return fieldArg.name === filter;
            });
          })) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected %s argument to @%s to be an array of ' + 'argument names on field `%s`, but get %s. Source: %s.', CLIENT_FIELD_FILTERS, CLIENT_FIELD, fieldName, maybeFilters, _this8._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
          // $FlowFixMe
          filters = maybeFilters.value;
        }
        handles = handles || [];
        handles.push({ name: _name2, key: key, filters: filters });
      }
    });
    return handles;
  };

  GraphQLParser.prototype._transformDirectives = function _transformDirectives(directives) {
    var _this9 = this;

    return directives.map(function (directive) {
      var name = getName(directive);
      var directiveDef = _this9._schema.getDirective(name);
      !directiveDef ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Unknown directive `@%s`. Source: %s.', name, _this9._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
      var args = _this9._transformArguments(directive.arguments || [], directiveDef.args);
      return {
        kind: 'Directive',
        metadata: null,
        name: name,
        args: args
      };
    });
  };

  GraphQLParser.prototype._transformArguments = function _transformArguments(args, argumentDefinitions) {
    var _this10 = this;

    return args.map(function (arg) {
      var argName = getName(arg);
      var argDef = argumentDefinitions.find(function (def) {
        return def.name === argName;
      });
      !argDef ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Unknown argument `%s`. Source: %s.', argName, _this10._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
      var value = _this10._transformValue(arg.value, argDef.type);
      return {
        kind: 'Argument',
        metadata: null,
        name: argName,
        value: value,
        type: argDef.type
      };
    });
  };

  GraphQLParser.prototype._splitConditions = function _splitConditions(mixedDirectives) {
    var _this11 = this;

    var conditions = [];
    var directives = [];
    mixedDirectives.forEach(function (directive) {
      if (directive.name === INCLUDE || directive.name === SKIP) {
        var passingValue = directive.name === INCLUDE;
        var arg = directive.args[0];
        !(arg && arg.name === IF) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected an `if` argument to @%s. Source: %s.', directive.name, _this11._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
        !(arg.value.kind === 'Variable' || arg.value.kind === 'Literal') ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected the `if` argument to @%s to be a variable. ' + 'Source: %s.', directive.name, _this11._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
        conditions.push({
          kind: 'Condition',
          condition: arg.value,
          metadata: null,
          passingValue: passingValue,
          selections: []
        });
      } else {
        directives.push(directive);
      }
    });
    var sortedConditions = [].concat(conditions).sort(function (a, b) {
      if (a.condition.kind === 'Variable' && b.condition.kind === 'Variable') {
        return a.condition.variableName < b.condition.variableName ? -1 : a.condition.variableName > b.condition.variableName ? 1 : 0;
      } else {
        // sort literals earlier, variables later
        return a.condition.kind === 'Variable' ? 1 : b.condition.kind === 'Variable' ? -1 : 0;
      }
    });
    return [sortedConditions, directives];
  };

  GraphQLParser.prototype._transformVariable = function _transformVariable(ast, type) {
    var variableName = getName(ast);
    this._recordAndVerifyVariableReference(variableName, type);
    return {
      kind: 'Variable',
      metadata: null,
      variableName: variableName,
      type: type
    };
  };

  /**
   * Transforms AST values into IR values, extracting the literal JS values of any
   * subtree of the AST that does not contain a variable.
   */


  GraphQLParser.prototype._transformValue = function _transformValue(ast, type) {
    var _this12 = this;

    switch (ast.kind) {
      case 'IntValue':
        return {
          kind: 'Literal',
          metadata: null,
          value: parseInt(ast.value, 10)
        };
      case 'FloatValue':
        return {
          kind: 'Literal',
          metadata: null,
          value: parseFloat(ast.value)
        };
      case 'StringValue':
        return {
          kind: 'Literal',
          metadata: null,
          value: ast.value
        };
      case 'BooleanValue':
        // Note: duplicated because Flow does not understand fall-through cases
        return {
          kind: 'Literal',
          metadata: null,
          value: ast.value
        };
      case 'EnumValue':
        // Note: duplicated because Flow does not understand fall-through cases
        return {
          kind: 'Literal',
          metadata: null,
          value: ast.value
        };
      case 'ListValue':
        var itemType = void 0;
        if (type) {
          var listType = require('./GraphQLSchemaUtils').getNullableType(type);
          // The user entered a list, a `type` was expected; this is only valid
          // if `type` is a List.
          !(listType instanceof require('graphql').GraphQLList) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected a value matching type `%s`, but ' + 'got a list value. Source: %s.', type, this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
          itemType = require('graphql').assertInputType(listType.ofType);
        }
        var literalList = [];
        var items = [];
        var areAllItemsScalar = true;
        ast.values.forEach(function (item) {
          var itemValue = _this12._transformValue(item, itemType);
          if (itemValue.kind === 'Literal') {
            literalList.push(itemValue.value);
          }
          items.push(itemValue);
          areAllItemsScalar = areAllItemsScalar && itemValue.kind === 'Literal';
        });
        if (areAllItemsScalar) {
          return {
            kind: 'Literal',
            metadata: null,
            value: literalList
          };
        } else {
          return {
            kind: 'ListValue',
            metadata: null,
            items: items
          };
        }
      case 'NullValue':
        return {
          kind: 'Literal',
          metadata: null,
          value: null
        };
      case 'ObjectValue':
        var literalObject = {};
        var fields = [];
        var areAllFieldsScalar = true;
        ast.fields.forEach(function (field) {
          var fieldName = getName(field);
          var fieldType = void 0;
          if (type) {
            var objectType = require('./GraphQLSchemaUtils').getNullableType(type);
            // The user entered an object, a `type` was expected; this is only
            // valid if `type` is an Object.
            !(objectType instanceof require('graphql').GraphQLInputObjectType) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected a value matching type `%s`, but ' + 'got an object value. Source: %s.', type, _this12._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
            var fieldConfig = objectType.getFields()[fieldName];
            !fieldConfig ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Unknown field `%s` on type `%s`. Source: %s.', fieldName, type, _this12._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
            fieldType = require('graphql').assertInputType(fieldConfig.type);
          }
          var fieldValue = _this12._transformValue(field.value, fieldType);
          if (fieldValue.kind === 'Literal') {
            literalObject[field.name.value] = fieldValue.value;
          }
          fields.push({
            kind: 'ObjectFieldValue',
            metadata: null,
            name: fieldName,
            value: fieldValue
          });
          areAllFieldsScalar = areAllFieldsScalar && fieldValue.kind === 'Literal';
        });
        if (areAllFieldsScalar) {
          return {
            kind: 'Literal',
            metadata: null,
            value: literalObject
          };
        } else {
          return {
            kind: 'ObjectValue',
            metadata: null,
            fields: fields
          };
        }
      case 'Variable':
        return this._transformVariable(ast, type);
      default:
        !false ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Unknown ast kind: %s. Source: %s.', ast.kind, this._getErrorContext()) : require('fbjs/lib/invariant')(false) : void 0;
    }
  };

  return GraphQLParser;
}();

function isScalarFieldType(type) {
  var namedType = require('graphql').getNamedType(type);
  return namedType instanceof require('graphql').GraphQLScalarType || namedType instanceof require('graphql').GraphQLEnumType;
}

function assertScalarFieldType(type) {
  !isScalarFieldType(type) ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'Expected %s to be a Scalar Field type.', type) : require('fbjs/lib/invariant')(false) : void 0;
  return type;
}

function applyConditions(conditions, selections) {
  var nextSelections = selections;
  conditions.forEach(function (condition) {
    nextSelections = [(0, _extends3['default'])({}, condition, {
      selections: nextSelections
    })];
  });
  return nextSelections;
}

function getName(ast) {
  var name = ast.name ? ast.name.value : null;
  !(typeof name === 'string') ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'GraphQLParser: Expected ast node `%s` to have a name.', ast) : require('fbjs/lib/invariant')(false) : void 0;
  return name;
}

/**
 * Partitions an array given a predicate. All elements satisfying the predicate
 * are part of the first returned array, and all elements that don't are in the
 * second.
 */
function partitionArray(array, predicate, context) {
  var first = [];
  var second = [];
  array.forEach(function (element, index) {
    if (predicate.call(context, element, index, array)) {
      first.push(element);
    } else {
      second.push(element);
    }
  });
  return [first, second];
}

module.exports = GraphQLParser;
